File: codeforces/[GuangDong]-25-12-4/I.go
package main

import (
	"bufio"
	. "fmt"
	"io"
	"os"
	"runtime/debug"
)

func lowerBound(nums *[]int, target int) int {
	left, right := -1, len(*nums)
	for left+1 != right {
		mid := (left + right) / 2
		if (*nums)[mid] >= target {
			right = mid
		} else {
			left = mid
		}
	}
	return right
}

func CF106210I(_in io.Reader, w io.Writer) {
	in := bufio.NewReader(_in)
	var n int
	Fscan(in, &n)
	nums := make([]int, n)
	pos := map[int][]int{}
	for i := 0; i < n; i++ {
		Fscan(in, &nums[i])
		pos[nums[i]] = append(pos[nums[i]], i)
	}
	ans := 0
	for i := 0; i < n; i++ {
		// 当前数字所有位置数组
		curPos := pos[nums[i]]
		posNum := lowerBound(&curPos, i)
		ans += n - (i + 1) - (len(curPos) - (posNum + 1))
	}
	Fprintln(w, ans)
}

func main() {
	debug.SetGCPercent(-1)
	CF106210I(os.Stdin, os.Stdout)
}

File: codeforces/[GuangDong]-25-12-4/I.cpp
#include <bits/stdc++.h>
using namespace std;
#define int long long
// 二分查找
int lowerBound(vector<int>& nums, int target) {
    int left = -1, right = nums.size();
    while (left + 1 != right) {
        int mid = (left + right) / 2;
        if (nums[mid] >= target) {
            right = mid;
        } else {
            left = mid;
        }
    }
    return right;
}
signed main() {
    cin.tie(0);
    cout.tie(0);
    ios::sync_with_stdio(false);
    int n;
    cin >> n;
    vector<int> nums(n);
    unordered_map<int, vector<int>> pos; // 这个将会记录对应的出现位置
    // 我将会采取二分查找方式
    for (int i = 0; i < n; ++i) {
        cin >> nums[i];
        pos[nums[i]].push_back(i);
    }
    int ans = 0;
    for (int i = 0; i < n; ++i) {
        auto& numPos = pos[nums[i]];
        auto thePosIter = lower_bound(numPos.begin(), numPos.end(), i);
        int thePos = thePosIter - numPos.begin();
        ans += n - (i + 1) - (numPos.size() - (thePos + 1));
    }
    cout << ans;
    return 0;
}